env0/resource_environment.go

```diff
@@ -550,20 +550,66 @@ func createVariable(configurationVariable *client.ConfigurationVariable) any {
 }
 
 // Validate that the template is assigned to the "project_id".
-func validateTemplateProjectAssignment(d *schema.ResourceData, apiClient client.ApiClientInterface) error {
+func validateTemplateProjectAssignment(d *schema.ResourceData, template *client.Template) error {
 	projectId := d.Get("project_id").(string)
+
+	if projectId != template.ProjectId && !stringInSlice(projectId, template.ProjectIds) {
+		return errors.New("could not create environment: template is not assigned to project")
+	}
+
+	return nil
+}
+
+func createEnvironmentWithTemplate(d *schema.ResourceData, apiClient client.ApiClientInterface) (client.Environment, client.EnvironmentCreate, diag.Diagnostics) {
 	templateId := d.Get("template_id").(string)
 
 	template, err := apiClient.Template(templateId)
 	if err != nil {
-		return fmt.Errorf("could not get template: %w", err)
+		return client.Environment{}, client.EnvironmentCreate{}, diag.Errorf("could not get template: %v", err)
 	}
 
-	if projectId != template.ProjectId && !stringInSlice(projectId, template.ProjectIds) {
-		return errors.New("could not create environment: template is not assigned to project")
+	if err := validateTemplateProjectAssignment(d, &template); err != nil {
+		return client.Environment{}, client.EnvironmentCreate{}, diag.Errorf("%v", err)
 	}
 
-	return nil
+	environmentPayload, diagError := getCreatePayload(d, apiClient, template.Type)
+	if diagError != nil {
+		return client.Environment{}, client.EnvironmentCreate{}, diagError
+	}
+
+	environment, err := apiClient.EnvironmentCreate(environmentPayload)
+	if err != nil {
+		return client.Environment{}, client.EnvironmentCreate{}, diag.Errorf("could not create environment: %v", err)
+	}
+
+	return environment, environmentPayload, nil
+}
+
+func createEnvironmentWithoutTemplate(d *schema.ResourceData, apiClient client.ApiClientInterface) (client.Environment, client.EnvironmentCreate, diag.Diagnostics) {
+	templatePayload, diagError := templateCreatePayloadFromParameters("without_template_settings.0", d)
+	if diagError != nil {
+		return client.Environment{}, client.EnvironmentCreate{}, diagError
+	}
+
+	environmentPayload, diagError := getCreatePayload(d, apiClient, templatePayload.Type)
+	if diagError != nil {
+		return client.Environment{}, client.EnvironmentCreate{}, diagError
+	}
+
+	payload := client.EnvironmentCreateWithoutTemplate{
+		EnvironmentCreate: environmentPayload,
+		TemplateCreate:    templatePayload,
+	}
+
+	// Note: the blueprint id field of the environment is returned only during creation of a template without environment.
+	// Afterward, it will be omitted from future response.
+	// setEnvironmentSchema() sets the blueprint id in the resource (under "without_template_settings.0.id").
+	environment, err := apiClient.EnvironmentCreateWithoutTemplate(payload)
```

Comment:
taste `createEnvironmentErr`

Reply:
The common/standard coding convention in golang is "err".
I will keep as is.